packages/nhsuk-frontend/src/nhsuk/components/file-upload/file-upload.mjs

```diff
@@ -0,0 +1,492 @@
+import { closestAttributeValue } from '../../common/closest-attribute-value.mjs'
+import { ConfigurableComponent } from '../../configurable-component.mjs'
+import { ElementError } from '../../errors/index.mjs'
+import { I18n } from '../../i18n.mjs'
+
+/**
+ * File upload component
+ *
+ * @preserve
+ * @augments ConfigurableComponent<FileUploadConfig>
+ */
+export class FileUpload extends ConfigurableComponent {
+  /**
+   * @private
+   * @type {HTMLFileInputElement}
+   */
+  $input
+
+  /**
+   * @private
+   */
+  $button
+
+  /**
+   * @private
+   */
+  $status
+
+  /** @private */
+  i18n
+
+  /** @private */
+  id
+
+  /** @private */
+  $announcements
+
+  /**
+   * @private
+   * @type {boolean | undefined}
+   */
+  enteredAnotherElement
+
+  /**
+   * @param {Element | null} $root - File input element
+   * @param {FileUploadConfig} [config] - File Upload config
+   */
+  constructor($root, config = {}) {
+    super($root, config)
+
+    const $input = this.$root.querySelector('input')
+
+    if ($input === null) {
+      throw new ElementError({
+        component: FileUpload,
+        identifier: 'File inputs (`<input type="file">`)'
+      })
+    }
+
+    if ($input.type !== 'file') {
+      throw new ElementError({
+        component: FileUpload,
+        identifier:
+          'File input (`<input type="file">`) attribute (`type`) is not `file`'
```

Comment:
Thanks for using `ElementError` here instead of `formatErrorMessage()`

Watch out though because `element` hasn't been passed in so it'll automatically append "not found" to the error message. How about something like this so `expectedType` can still be used?

```suggestion
        element: $input,
        identifier: 'Form field (`<input>`)',
        expectedType: 'HTMLInputElement with type="file"',
```

Reply:
Or keeping the expected attribute in backticks?

```suggestion
        element: $input,
        identifier: 'Form field (`<input>`)',
        expectedType: 'HTMLInputElement with attribute (`type="file"`)',
```

Reply:
Fixed.